File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();
        int R=12, L=10;
        for(int i=0;i<numbers.length;i++){
            int n=numbers[i];
            if(n==1||n==4||n==7){
                L=n;
                answer.append("L");
            }    
            else if(n==3||n==6||n==9){
                R=n;
                answer.append("R");
            }   
            else{
                if(n==0)    n=11;
                int lm = Math.abs(n-L)/3+Math.abs(n-L)%3;
                int rm = Math.abs(n-R)/3+Math.abs(n-R)%3;
                if(lm==rm){
                    if(hand.equals("right")){
                        R=n;
                        answer.append("R");
                    }  
                    else if(hand.equals("left")){
                        L=n;
                        answer.append("L");
                    }
                }
                else if(lm<rm){
                    L=n;
                    answer.append("L"); 
                }  
                else{
                    R=n;
                    answer.append("R");
                }    
            }
        }
        return answer.toString();
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    l,r=10,12
    
    for n in numbers:
        if n==0:    n=11
        if n in [1,4,7]:
            answer+='L'
            l=n
        elif n in [3,6,9]:
            answer+='R'
            r=n
        else:
            lmv=abs((l+1 if l in [1,4,7,10] else l)-n)//3+(1 if l in[1,4,7,10] else 0)
            rmv=abs((r-1 if r in [3,6,9,12] else r)-n)//3+(1 if r in[3,6,9,12] else 0)
            # print("l:",l,"r:",r,"n:",n)
            # print("lmv:",lmv,"rmv:",rmv)
            if lmv==rmv: 
                if hand=='right':
                    answer+='R'
                    r=n
                else:
                    answer+='L'
                    l=n
            else:
                if rmv<lmv:
                    answer+='R'
                    r=n
                else:
                    answer+='L'
                    l=n

    return answer